File: src/dog.ts
import { Animal } from "@/animal";

export interface Dog extends Animal {}

export class Dog implements Dog {
  private _name: string;
  private _age: number;

  constructor(parameters: { name: string; age: number }) {
    this._name = parameters.name;
    this._age = parameters.age;
  }

  public show(): void {
    console.log(`Hi, i'm ${this._name} and i'm ${this._age} yo!`);
  }

  public sound(): void {
    console.log("Woof!");
  }

  public toString(): string {
    return `Dog { name: ${this._name}, age: ${this._age} }`;
  }
}

File: src/dog.rs
use crate::animal::Animal;

pub struct Dog {
    name: String,
    age: u32,
}

impl Dog {
    pub fn new(name: String, age: u32) -> Self {
        Dog { name, age }
    }
}

impl Animal for Dog {
    fn show(&self) -> () {
        println!("Hi, I'm {}!, I'm {} years old!", self.name, self.age);
    }

    fn sound(&self) -> () {
        println!("Woof!");
    }

    fn name(&self) -> &str {
        &self.name
    }

    fn age(&self) -> u32 {
        self.age
    }

    fn to_string(&self) -> String {
        format!("Dog[name={}, age={}]", self.name, self.age)
    }
}
